### PR TITLE
gh-106706: Streamline family syntax

### PR DESCRIPTION
**Pull request by kgdiem**
* Update bytecodes.c to use suggested syntax from #106706 
* Update checks in `generate_cases.py` to support >= 1 members and validate family name 
* Update documentation 
* Update tests 

* Issue: gh-106706

### CODE REVIEW

File: Tools/cases_generator/test_generator.py [hunk 1]
```diff
@@ -321,13 +321,13 @@ def test_macro_instruction():
                 _tmp_3 = res;
             }
             next_instr += 5;
-            static_assert(INLINE_CACHE_ENTRIES_OP == 5, "incorrect cache size");
```
Comment (kgdiem) [inline Tools/cases_generator/test_generator.py]: It isn't clear to me how/where the INLINE_CACHE_ENTRIES_OP instruction is generated. 

Changing the test to pass w/o understanding doesn't inspire confidence on my end so any guidance or explanation here is greatly appreciated.

**Comment by kgdiem** [comment]
Also

```
➜  cpython git:(main) ✗ python3 ./Tools/cases_generator/generate_cases.py 
Read 210 instructions/ops, 17 macros, 12 pseudos, and 13 families from Python/bytecodes.c
Python/bytecodes.c:285: Family 'TO_BOOL' has unknown instruction 'TO_BOOL'
Python/bytecodes.c:372: Family 'BINARY_OP' has unknown instruction 'BINARY_OP'
Python/bytecodes.c:506: Family 'BINARY_SUBSCR' has unknown instruction 'BINARY_SUBSCR'
Python/bytecodes.c:641: Family 'STORE_SUBSCR' has unknown instruction 'STORE_SUBSCR'
Python/bytecodes.c:918: Family 'SEND' has unknown instruction 'SEND'
Python/bytecodes.c:1130: Family 'UNPACK_SEQUENCE' has unknown instruction 'UNPACK_SEQUENCE'
Python/bytecodes.c:1193: Family 'STORE_ATTR' has unknown instruction 'STORE_ATTR'
Python/bytecodes.c:1292: Family 'LOAD_GLOBAL' has unknown instruction 'LOAD_GLOBAL'
Python/bytecodes.c:1640: Family 'LOAD_SUPER_ATTR' has unknown instruction 'LOAD_SUPER_ATTR'
Python/bytecodes.c:1742: Family 'LOAD_ATTR' has unknown instruction 'LOAD_ATTR'
Python/bytecodes.c:2026: Family 'COMPARE_OP' has unknown instruction 'COMPARE_OP'
Python/bytecodes.c:2327: Family 'FOR_ITER' has unknown instruction 'FOR_ITER'
Python/bytecodes.c:2717: Family 'CALL' has unknown instruction 'CALL'
Found 13 errors
```

File: Misc/NEWS.d/next/Tools-Demos/2023-07-13-12-08-35.gh-issue-106706.29zp8E.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Change bytecode syntax for families
+to remove redundant name matching
+pseudo snytax
```
Comment (gvanrossum) [inline Misc/NEWS.d/next/Tools-Demos/2023-07-13-12-08-35.gh-issue-106706.29zp8E.rst]:
```suggestion
pseudo syntax.
```

File: Python/executor_cases.c.h [hunk 3]
```diff
@@ -183,6 +182,7 @@
         }
 
         case TO_BOOL_ALWAYS_TRUE: {
+            static_assert(INLINE_CACHE_ENTRIES_TO_BOOL == 3, "incorrect cache size");
```
Comment (gvanrossum) [inline Python/executor_cases.c.h]: While it works, I think it's better to keep the static assert in the "family head".

File: Tools/cases_generator/generate_cases.py [hunk 4]
```diff
@@ -865,8 +865,13 @@ def check_families(self) -> None:
         - All members must have the same cache, input and output effects
         """
         for family in self.families.values():
-            if len(family.members) < 2:
+            if len(family.members) < 1:
```
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py]: I don't think this can even happen any more -- if you try to create an empty family it will be a SyntaxError.

File: Tools/cases_generator/generate_cases.py [hunk 5]
```diff
@@ -877,10 +882,10 @@ def check_families(self) -> None:
                 self.error(
                     f"Family {family.name!r} has unknown members: {unknown}", family
                 )
-            if len(members) < 2:
+            if len(members) < 1:
```
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py]: Ditto.

File: Tools/cases_generator/generate_cases.py [hunk 6]
```diff
@@ -1303,11 +1308,11 @@ def write_metadata(self) -> None:
             self.out.emit("")
             self.out.emit("_specializations = {")
             for name, family in self.families.items():
-                assert len(family.members) > 1
+                assert len(family.members) >= 1
```
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py]: Ditto

File: Tools/cases_generator/interpreter_definition.md [hunk 7]
```diff
@@ -351,7 +351,7 @@ A _family_ represents a specializable instruction and its specializations.
```
Comment (gvanrossum) [inline Tools/cases_generator/interpreter_definition.md:351]: Perhaps the line above could say "A familiy *maps* a specializable instruction *to* ..." ?

File: Tools/cases_generator/test_generator.py (hunk 1 above)
Comment (gvanrossum) [reply Tools/cases_generator/test_generator.py]: It is the second parameter to `family`.

**Review by gvanrossum — COMMENTED** [review]
Thank you! I hope I've given you enough hints on what to do about the static assert -- if not, let me know.

**Comment by kgdiem** [comment]
Almost done, just need to dig in a bit further to see where the size is being written / update generator to put the size in the head block.

File: .coveragerc [hunk 8]
```diff
@@ -0,0 +1,19 @@
+[run]
```
Comment (gvanrossum) [inline .coveragerc]: This file doesn't belong in this PR.

File: Lib/_opcode_metadata.py [hunk 9]
```diff
@@ -31,75 +31,3 @@
         "STORE_SUBSCR_DICT",
         "STORE_SUBSCR_LIST_INT",
     ],
-    "SEND": [
```
Comment (gvanrossum) [inline Lib/_opcode_metadata.py]: That this file changed surprises me. Maybe this points to a problem?

File: Python/executor_cases.c.h (hunk 3 above)
Comment (gvanrossum) [reply Python/executor_cases.c.h]: This still isn't resolved?

File: Tools/cases_generator/generate_cases.py [hunk 10]
```diff
@@ -855,8 +855,11 @@ def check_families(self) -> None:
         - All members must have the same cache, input and output effects
         """
         for family in self.families.values():
-            if len(family.members) < 2:
-                self.error(f"Family {family.name!r} has insufficient members", family)
+            if family.name not in self.macro_instrs and family.name not in self.instrs:
+                self.error(
+                    f"Family {family.name!r} has unknown instruction {family.name!r}",
+                    family,
+                )
```
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py:862]: Good catch!

**Review by gvanrossum — COMMENTED** [review]
I think you pushed a rebase? Please don't do that once review is in progress -- everything will be squashed when we merge into main.

Can you look into moving the static_assert back to the family head?

And I'm unsure why the metadata file was truncated. (Let me know if you need help figuring this out.)

**Comment by kgdiem** [comment]
> 

Ya there was a conflict in one of the header files so I rebased and re-generated it. Will keep that in mind next time.

**Comment by kgdiem** [comment]
> I think you pushed a rebase? Please don't do that once review is in progress -- everything will be squashed when we merge into main.
> 
> Can you look into moving the static_assert back to the family head?
> 
> And I'm unsure why the metadata file was truncated. (Let me know if you need help figuring this out.)

Ya there was a conflict in one of the header files so I rebased and re-generated it. Will keep that in mind next time.

File: Lib/_opcode_metadata.py (hunk 9 above)
Comment (kgdiem) [reply Lib/_opcode_metadata.py]: It didn't change originally, I think only after https://github.com/python/cpython/pull/106716/commits/178e3009234e49102d2d8058504f9c25eb374a06 ?

File: Python/executor_cases.c.h (hunk 3 above)
Comment (kgdiem) [reply Python/executor_cases.c.h]: No, I think this might be related to the change in `Lib/_opcode_metadata.py`as well; the op from the family name isn't getting inserted into `instrs`

I'm having some trouble getting the name into `instrs` / parsed as an instruction. Any advice?

**Comment by gvanrossum** [comment]
> Ya there was a conflict in one of the header files so I rebased and re-generated it. Will keep that in mind next time.

The missing part of the metadata will be reconstructed if you simply run `make regen-cases` and then commit the resulting changes to Lib/_opcode_metadata.py.

Not sure yet why the stat asserts moved, will look into it later. (Did you find where they are generated?)

**Comment by kgdiem** [comment]
> > Ya there was a conflict in one of the header files so I rebased and re-generated it. Will keep that in mind next time.
> 
> The missing part of the metadata will be reconstructed if you simply run `make regen-cases` and then commit the resulting changes to Lib/_opcode_metadata.py.
> 
> Not sure yet why the stat asserts moved, will look into it later. (Did you find where they are generated?)

Yes, https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L1555 

This condition is always falsy because `last_instr.family` is `None` on the instruction used as `family.name`. 

It misses in `map_families` (https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L835)  because it isn't added to the `instrs` in `parse_file` https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L787

**Comment by gvanrossum** [comment]
> I'm thinking that there would be far fewer changes to `generate_cases` while keeping the desired syntax for a family template if the family name was added to the members here https://github.com/python/cpython/blob/main/Tools/cases_generator/parser.py#L347
> 
> eg `..., [tkn.txt] + members`
> 
> I considered just starting fresh from there but thought I would check given we've been back and forth on this implementation for a couple days.

Interesting idea. It would mean fewer changes, but some of the ugliness would remain as well (e.g. the several places where you have to slice `.members[1:]`).

As you point out, the problem is that the family head doesn't have `self.family` set to the family whose head it is. I think this will fix it:
```diff
diff --git a/Tools/cases_generator/generate_cases.py b/Tools/cases_generator/generate_cases.py
index 48c082a429..4756f50b8f 100644
--- a/Tools/cases_generator/generate_cases.py
+++ b/Tools/cases_generator/generate_cases.py
@@ -438,7 +438,7 @@ def write(self, out: Formatter, tier: Tiers = TIER_ONE) -> None:
         """Write one instruction, sans prologue and epilogue."""
         # Write a static assertion that a family's cache size is correct
         if family := self.family:
-            if self.name == family.members[0]:
+            if self.name == family.name:
                 if cache_size := family.size:
                     out.emit(
                         f"static_assert({cache_size} == "
@@ -831,7 +831,7 @@ def find_predictions(self) -> None:
     def map_families(self) -> None:
         """Link instruction names back to their family, if they have one."""
         for family in self.families.values():
-            for member in family.members:
+            for member in [family.name] + family.members:
                 if member_instr := self.instrs.get(member):
                     if member_instr.family not in (family, None):
                         self.error(
```

**Comment by kgdiem** [comment]
Awesome! I’ll get everything cleaned up in a little bit.

**Comment by kgdiem** [comment]
> This condition is always falsy because `last_instr.family` is `None` on the instruction used as `family.name`.
> 
> That instruction misses in `map_families` (https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L835) because it isn't added to the `instrs` in `parse_file` https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L787

I committed those changes but I'd actually outlined these changes in my previous comment:

>This condition is always falsy because last_instr.family is None on the instruction used as family.name.
>
>That instruction misses in map_families (https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L835) because it isn't added to the instrs in parse_file https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L787

`family.name` is never mapped because it isn't present in `self.instrs`; This results in the assertion never being emitted.

**Comment by kgdiem** [comment]
> > This condition is always falsy because `last_instr.family` is `None` on the instruction used as `family.name`.
> > That instruction misses in `map_families` (https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L835) because it isn't added to the `instrs` in `parse_file` https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L787
> 
> I committed those changes but I'd actually outlined them in my previous comment:
> 
> > This condition is always falsy because last_instr.family is None on the instruction used as family.name.
> > That instruction misses in map_families (https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L835) because it isn't added to the instrs in parse_file https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L787
> 
> `family.name` is never mapped because it isn't present in `self.instrs`; This results in the assertion never being emitted.

Whelp, looks like that's actually just because of how this test was written and it fixes the headers. My bad! Fixing test.

**Comment by kgdiem** [comment]
> > > This condition is always falsy because `last_instr.family` is `None` on the instruction used as `family.name`.
> > > That instruction misses in `map_families` (https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L835) because it isn't added to the `instrs` in `parse_file` https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L787
> > 
> > 
> > I committed those changes but I'd actually outlined them in my previous comment:
> > > This condition is always falsy because last_instr.family is None on the instruction used as family.name.
> > > That instruction misses in map_families (https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L835) because it isn't added to the instrs in parse_file https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L787
> > 
> > 
> > `family.name` is never mapped because it isn't present in `self.instrs`; This results in the assertion never being emitted.
> 
> Looks like the generated headers are now correct / unchanged and the test is testing for a macro instruction and what I described above is still missing here https://github.com/python/cpython/blob/main/Tools/cases_generator/generate_cases.py#L1555
> 
> But the generated header files after that change are correct.
> 
> Makes a lot more sense now why the macro isn't in `self.instrs`, wasn't thinking about it correctly before. Still unsure of what to do or if the test is wrong.

Got it! Instead of checking the last instruction, get the family from `self.families`

**Review by gvanrossum — APPROVED** [review]
Perfect! I'll merge.
